docs: :memo: add exercises (#24)
rostools/tutorial-github-intro: sessions/lesson.qmd

TODO: Reading text
:::

## Exercise: Add a new file and modify the content

First, we want to add a file to the repository.

1.  Go to the repository on GitHub.
2.  In the code tab, click "Add file" and then "Create new file".
3.  In the "Name your file..." field, type a name for your file. For
    simplicity, let's call it your name and make it a Markdown file. For
    example, "your-name.md".
4.  Click the green "Commit changes" button and fill in the commit
    message. For example, "Add file with my name". You don't need to
    fill in the extended description. Keep the "Commit directly to the
    `main` branch" option selected.
5.  When you have filled in the commit message, click the green "Commit
    changes" button.
6.  Go back to the repository and check that your file has been added.

> You have now made your first commit to the repository! :tada:

Author: I guess this could be a part of the code along instead?

Reviewer: Yea, these steps could be, but the basic steps can stay the same. See comment above about "problem to solve"

Author: For now, I have kept everything as exercises. Then we can move them to the different lessons and change them accordingly there.

3.  Make a change to the file by adding a new line of text. For example,
    "This is a new line of text."
4.  Now, commit the changes as you did before by clicking the green
    "Commit changes" button and filling in the commit message. When you
    have filled in the commit message, click the green "Commit changes"
    button.
5.  Go back to the file on GitHub and check that your changes have been
    made.

::: callout-tip
## See the history of changes

If you want to see the history of changes to the file, go to the file
and click the "History" button on the right side of the screen. This
will show you all the changes that have been made to the file along with
the commit messages.

Since you have made two commits to the file, you should see two entries
in the history. You can click on each entry to see the changes that were
made in that commit.

This also shows the importance of meaningful commit messages. Imagine if
all the commit messages were "Update file.md". Then, it would be hard to
know what changes were made in each commit. By contrast, if the commits
have meaningful messages, it would be much more easy to get an overview
of the changes that have been made.

> :::

Author: I thought this was a nice addition - so they try actually looking at the history they made - but I think this exercise became quite/too long?

Reviewer: Instead of a tip, make it an exercise! :grin: 

And more exercises are fine, it's easy to remove than it is to grow, and we can always put them to the side/use as extras.

{{< text_snippet sticky_up >}}

## Exercise: Rename and move your file

Now it's time to make edits to the file itself. We want to rename the
file and move it to a new location in the repository.

1.  Go to the file on GitHub.
2.  Click the pencil icon on the right side of the screen ("Edit this
    file") to edit the file.
3.  Now, click the name of the file at the top of the screen. After the
    current name of the file, add "-new-filename". For example, if the
    current name of the file is "file.txt", change it to
    "file-new-filename.txt". Remember to keep the file extension (e.g.,
    ".txt") the same.
4.  Commit the changes as we have done it previously, by clicking the
    green "Commit changes" button and fill in the commit message.
5.  Go back to the file on GitHub and check that your changes have been

>     made.

Author: Do we want the deletion of the file to be a part of the exercise as well?

{{< text_snippet sticky_up >}}

## Exercise: Create an issue and interact with another's issue

Let's create your first issue!

1.  In the repository on GitHub, click the "Issues" tab.
2.  Click the green "New issue" button.
3.  Fill in the title and the description. This could be anything you
    want to work on or discuss. For example, it could be about something
    you think should be added (a new file or a new section to a file) or
    something you think should be changed. It could also be a discussion
    issue where you ask for feedback on something.
4.  Click the green "Create" button, to create the issue.

Now, you have created your first issue! :partying_face:

Now, let's interact with another's issue:

1.  Go back to the "Issues" tab in the repository.
2.  Click on an issue that someone else has created.
3.  Read the issue and the comments that have been made.
4.  Scroll down to the bottom of the page and write a comment in the
    comment field. You can ask a question, give feedback, or suggest a
    solution to the issue.
5.  When you have written your comment, click the green "Comment" button
    to post the comment.

{{< text_snippet sticky_up >}}

> ## Exercise: Close your issue

Author: I made this a separate issue so the fastest wouldn't start closing their issues before the rest had time to comment/interact.

Reviewer: We could join the closing issue and delete file into one exercise called "cleaning up" or something.

TODO: Reading text
:::

## Exercise: Add a new file and modify the content

> First, we want to add a file to the repository.

Reviewer: Follow the format of the vscode snippet `exercise` to match the other websites :relaxed:

Reviewer: It helps that exercises have some problem to solve rather than "do this", "but why", "because". Maybe it could be like the exercises you thought of for git4cats, where we get them to find a recipe they like and put it on the repo as a new file.

Author: I've gone all out with the recipes / recipe book now :cake: :cookie:

    simplicity, let's call it your name and make it a Markdown file. For
    example, "your-name.md".
4.  Click the green "Commit changes" button and fill in the commit
    message. For example, "Add file with my name". You don't need to
    fill in the extended description. Keep the "Commit directly to the
    `main` branch" option selected.
5.  When you have filled in the commit message, click the green "Commit
    changes" button.
6.  Go back to the repository and check that your file has been added.

You have now made your first commit to the repository! :tada:

Now that we have created a file, let's modify it:

1.  Go to the file on GitHub
2.  Click the pencil icon on the right side of the screen ("Edit this
    file") to edit the file
3.  Make a change to the file by adding a new line of text. For example,
    "This is a new line of text."
4.  Now, commit the changes as you did before by clicking the green
    "Commit changes" button and filling in the commit message. When you
    have filled in the commit message, click the green "Commit changes"
    button.
5.  Go back to the file on GitHub and check that your changes have been

>     made.

Reviewer: Based on my experience and the feedback I've gotten, multiple part exercises don't work too well. Keep them smaller and shorter for intro sessions. Or rather, keep the exercise more targeted in scope. This exercise is doing several things.

::: callout-tip
## See the history of changes

If you want to see the history of changes to the file, go to the file
and click the "History" button on the right side of the screen. This
will show you all the changes that have been made to the file along with
the commit messages.

Since you have made two commits to the file, you should see two entries
in the history. You can click on each entry to see the changes that were
made in that commit.

This also shows the importance of meaningful commit messages. Imagine if
all the commit messages were "Update file.md". Then, it would be hard to
know what changes were made in each commit. By contrast, if the commits
have meaningful messages, it would be much more easy to get an overview
of the changes that have been made.
:::

{{< text_snippet sticky_up >}}

## Exercise: Rename and move your file, then delete it

Now it's time to make edits to the file itself. We want to rename the

> file and move it to a new location in the repository.

Reviewer: See comment above about "problem to solve"

    current name of the file, add "-new-filename". For example, if the
    current name of the file is "file.md", change it to
    "your-name-new-filename.md". Remember to keep the file extension
    (e.g., ".md") the same.
4.  Commit the changes as we have done it previously, by clicking the
    green "Commit changes" button and fill in the commit message.
5.  Go back to the file on GitHub and see that your changes have been
    made.

Now, let's move the file to a new folder in the repository:

1.  Go to the file.
2.  Click the pencil icon on the right side of the screen ("Edit this
    file") as we did before.
3.  Click the name of the file at the top of the screen. Change the path
    of the file to a new location in the repository. For example, if the
    current path is "your-name-new-filename.md", change it to
    "folder/your-name-new-filename.md". Notice the `/` between the
    folder name and the file name.
4.  Commit the changes.

Lastly, let's delete the file:

1.  Go to the file.
2.  On the same line as the file name, click the button with the three
    dots, saying "More file actions". Click it, scroll down and click
    "Delete file".

> 3.  Commit the changes.

Reviewer: See comment below, this could be moved down.

own and the changes made by others).
:::

{{< text_snippet sticky_up >}}

## Exercise: Create an issue and interact with another's issue

Let's create your first issue!

1.  In the repository on GitHub, click the "Issues" tab.
2.  Click the green "New issue" button.
3.  Fill in the title and the description. This could be anything you
    want to work on or discuss. For example, it could be about something
    you think should be added (a new file or a new section to a file) or
    something you think should be changed. It could also be a discussion
    issue where you ask for feedback on something.
4.  Click the green "Create" button, to create the issue.

Now, you have created your first issue! :partying_face:

Now, let's interact with another's issue:

1.  Go back to the "Issues" tab in the repository.
2.  Click on an issue that someone else has created.
3.  Read the issue and the comments that have been made.
4.  Scroll down to the bottom of the page and write a comment in the
    comment field. You can ask a question, give feedback, or suggest a
    solution to the issue.
5.  When you have written your comment, click the green "Comment" button

>     to post the comment.

Reviewer: This could be a code-along, with the "cleaning up" exercise being the end.

TODO: Reading text
:::

> ## :woman_technologist: Exercise: Add a cookie recipe to the repository {#sec-add-recipe-file}

Author: Also only added it bc I referenced them below. But I see that you've removed that

> Time: \~10 minutes.

Let's add the first recipe to your repository: A cookie recipe! :cookie:

First, we want to add a file that will include the recipe to the
repository.

1.  Go to the repository on GitHub.
2.  In the code tab, click "Add file" and then "Create new file".
3.  In the "Name your file..." field, type a name for your file. Let's
    call it "cooke-recipe.md". Remember the ".md" file extension so we
    create a [Markdown](https://www.markdownguide.org/getting-started/)
    file.
4.  Click the green "Commit changes" button and fill in the commit
    message. For example, "Add file with cooke recipe". You don't need
    to fill in the extended description. Keep the "Commit directly to
    the `main` branch" option selected.
5.  When you have filled in the commit message, click the green "Commit
    changes" button.
6.  Go back to the repository and check that your file has been added.

You have now added your first recipe and made your first commit to your
repository! :tada:

{{< text_snippet sticky_up >}}

## :technologist: Exercise: Add the cookie recipe to `cookie-recipe.md` file {#sec-add-recipe-content}

> > Time: \~10 minutes.

Author: Should this be changed in the snippet then?

Now that we have created the `cookie-recipe.md` file, let's modify it by
adding the recipe to it:

1.  Go to the file `cookie-recipe.md` in your GitHub repository

2.  Click the pencil icon on the right side of the screen ("Edit this
    file") to edit the file.

3.  Make a change to the file by adding a cookie recipe. If you (like
    most of us) don't know a cookie recipe by heart, you can find one
    online, or use the example recipe below (inspired by

>     [this](https://twin-food.dk/mit-bud-pa-de-bedste-cookies/) recipe -

Author: True..... We could also just say that they should add this cookie recipe to make sure everyone has the same files?

    file to `chocolate-chip-cookies.md`. Remember to avoid spaces in the
    file name and keep the file extension (`.md`).
4.  Commit the changes as we have done it previously, by clicking the
    green "Commit changes" button and fill in the commit message.
5.  Go back to the file on GitHub and see that your changes have been
    made.

Now, let's move the file to a new `cookies` folder in the repository:

1.  Go to the file.
2.  Click the pencil icon on the right side of the screen ("Edit this
    file") as we did before.
3.  Click the name of the file at the top of the screen. Change the path
    of the file to a new location in the repository. The current path is
    `chocolate-chip-cookies.md`. To create a new folder, change the path
    to `cookies/chocolate-chip-cookies.md`. Notice the `/` between the
    folder name and the file name.
4.  Commit the changes.

You have now renamed your recipe file and moved it to a new folder in
the repository! :partying_face:

{{< text_snippet sticky_up >}}

## :technologist: Exercise: Create an issue with a list of future recipes for your book

> Time: \~10 minutes.

> Now that you have added a recipe to the repository, let's create a
> GitHub Issue to keep track of the recipes you want to add to your recipe
> book in the future.

Reviewer: ```suggestion
You may or may not want to make a to-do list of other recipes to
add to the repository. Assuming you do want to, you can use GitHub
Issues to keep track of the recipes you want to add. Time to make one!
```

> 1.  In the repository on GitHub, click the "Issues" tab.

Reviewer: ```suggestion
1.  In your repository, click the "Issues" tab.
```

2.  Click the green "New issue" button.

> 3.  Fill in the title and the description. For example, you could write

Reviewer: ```suggestion
3.  Fill in the title with text like "List of future recipes".
    Then fill in the description box with some recipes you
    would like to add. If you don't know any recipes or are 
    not sure, you could write:
```

>     "List of future recipes" as the title and "I want to add the
>     following recipes to the recipe book:
>     -   Pancakes
>     -   Apple pie" as the description.

Reviewer: ````suggestion
    
    ``` markdown
    I want to add the following recipes to the recipe book:
    
    -   Pancakes
    -   Apple pie
    ```
````

> 4.  Click the green "Create" button, to create the issue.

Reviewer: ```suggestion
4.  Click the green "Create" button to create the issue.
```

Danish comma :stuck_out_tongue:

Author: Not even! Just a regular error in both Danish and English 😛

> Now, you have created your first issue! :tada:

Reviewer: ```suggestion
You now have an issue to remind yourself of recipes to add! :tada:
```

{{< text_snippet sticky_up >}}

> ## :woman_technologist: Exercise: Add a recipe suggestion to another's issue

Reviewer: ```suggestion
## :woman_technologist: Exercise: Comment with a recipe suggestion to your neighbour's issue
```

> > Time: \~10 minutes.

Reviewer: ```suggestion
**Time: \~10 minutes**
```

> Since one of the nice things about GitHub is that it allows for easy
> collaboration, let's interact with another's issue by adding a recipe
> suggestion to it. With your neighbour:

Reviewer: ```suggestion
One of the great things about GitHub is that it makes it easier to
collaborate with others. So, interact with someone else's issue by adding a comment to it:
```

> 1.  Go to each other's GitHub repositories.

Reviewer: ```suggestion
1.  One at a time, go to your neighbour's GitHub repository.
```

> 2.  Go to the "Issues" tab in the repository.

Reviewer: ```suggestion
2.  Go to the "Issues" tab in their repository.
```

3.  Click on the issue that they have created with the list of future
    recipes.
4.  Scroll down to the bottom of the page and write a comment in the
    comment field suggesting a recipe that they could add to their

>     recipe book. For example, you could suggest them to add a brownie

Reviewer: ```suggestion
    recipe book. For example, you could suggest that they add a brownie
```

    recipe as well :cake:.
5.  When you have written your comment, click the green "Comment" button
    to post the comment.

{{< text_snippet sticky_up >}}

> ## :technologist: Exercise: Add a recipe to another's repository :cake:

Reviewer: ```suggestion
## :technologist: Exercise: Add a recipe to your neighbour's repository :cake:
```

I'm seeing a pattern with the use of `another` :stuck_out_tongue: I guess the direct Danish translation for "andre" is "another", but in English it needs another linking word (like "another person's"). The more accurate translation would be "someone else's"

Author: Truuueee... Woopsi!

> Time: \~10 minutes.

::: {.callout-note collapse="true"}
## :teacher: Instructor note

Before starting this exercise, make sure that the learners have added
their neighbour as collaborators to their repositories. If they haven't,
they can do so by going to the repository, clicking on "Settings", then
"Collaborators and teams", and then "Add people".

> :::

Reviewer: This could be a task in the exercise, depending on how the content before is structured. For now, this is good!

Author: Agreed. Just wanted to add it, so we don't forget it.

Now, let's add a recipe to your neighbour's repository.

1.  Go to your neighbour's GitHub repository.

> 2.  As we did in @sec-add-recipe-file and @sec-add-recipe-content , add
>     a new file with a recipe - this time, it's just in another person's
>     repository. For example, you could add a brownie recipe like the one
>     below:

Reviewer: ```suggestion
2.  As we've done during this workshop, add
    a new file with a recipe, naming the file by using the recipe name (remember the correct naming for files!). 
    This time, you will be adding the file to your 
    neighbour's repository. In the file, add a recipe of your choice.
    For convenience, you can also use the brownie recipe below:
```

    ``` markdown
    # Brownies

    ## Ingredients 

    - 350 grams dark chocolate
    - 250 grams butter
    - 4 eggs
    - 150 grams brown sugar
    - 150 grams white sugar
    - 60 grams flour
    - 1 tsp baking powder

    ## Instructions

    1. Melt the chocolate and butter together.
    2. Whisk the eggs until frothy and add brown and white sugar.
    3. Add the melted chocolate and butter to the egg mixture.
    4. Add the flour and baking powder.
    5. Pour the mixture into a baking tray and bake at 170°C for 30-35 minutes.
    ```

> 3.  Commit the changes as we have done it previously and go to the

Reviewer: ```suggestion
3.  Commit the changes as you have done previously in the workshop. Then go to the neighbour's
```

    repository's history to see the changes you have made.

{{< text_snippet sticky_up >}}

> ## :woman_technologist: Exercise: See the changes in the repository
> 
> > Time: \~10 minutes.
> 
> After you have made the changes, go back to the repository and click the
> History button to see the changes that have been made (including your
> own and the changes made by others).
> 
> {{< text_snippet sticky_up >}}

Reviewer: Hmm, not sure this is necessary as an exercise. Unless it is re-written to be about a problem/finding something. We already have lots of exercises, so I think it's fine to delete.

```suggestion
```

> ## :woman_technologist: Exercise: Cleaning up
> 
> > Time: \~10 minutes.
> 
> Now, let's close the "Future recipes" issue you created earlier.
> 
> 1.  Go to the "Issues" tab in your repository.
> 2.  Click on the issue you created.
> 3.  In the comment field, write a comment about why you are closing the
>     issue. For example, you could write "This issue has been resolved"
>     or "I have now added all the relevant recipes from this issue".
>     Then, click the "Close issue" button.
> 
> Lastly, let's delete the `chocolate-chip-cookies.md` file along with the
> `cookies` folder you created earlier:
> 
> 1.  Go to the file.
> 2.  On the same line as the file name, click the button with the three
>     dots, saying "More file actions". Click it, scroll down and click
>     "Delete file".
> 3.  Commit the changes.
> 4.  Now, go the `cookies` folder and delete it in the same way as you
>     deleted the `chocolate-chip-cookies.md` file. For folders, it will
>     say "Delete directory" instead of "Delete file".
> 
> {{< text_snippet sticky_up >}}

Reviewer: ```suggestion
```

Given we already have lots of exercises, and that this is a low priority/low value exercise, let's just delete it.
